### PR TITLE
AirfRANS: vol-weight=15x & 20x clean control (no extra features)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Higher volume loss weighting alone (15x and 20x) on the confirmed champion config may push vol_mse below 0.002039 without regressing surface_mse. This is a clean control experiment — explicitly WITHOUT asinh-pressure or residual-prediction features (nezuko #3177 tests those same weights WITH those features).

**New directive:** AF work must focus on volume error reduction only without regressing surface_mse. External target: vol_mse < 0.0017 (currently 0.002039, 1.20x gap).

## Instructions

Mandatory champion config: AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, 2L/256d, **4 heads** (NOT 8).

**Run a 3-epoch smoke test on Trial 1 first.** Report BOTH surface_mse AND vol_mse before proceeding to full training.

**Trial 1 — vol-weight=15x clean:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 15.0 \
  --wandb_group af-vol-weight-higher
```

**Trial 2 — vol-weight=20x clean:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 20.0 \
  --wandb_group af-vol-weight-higher
```

Do NOT add `--asinh-pressure`, `--residual-prediction`, or any other noam features. This is an isolation experiment.

Report best surface_mse and vol_mse for each trial (they may peak at different epochs — report both epoch numbers).

## Baseline

| Metric | Value | Epoch |
|---|---|---|
| `val_primary/surface_mse` | **0.000296** | 704 |
| `val_primary/vol_mse` | **0.002039** | 743 |

- Baseline W&B run: `sh2zyfwr` (PR #3135 — EMA=0.999 + vol-weight=10x)
- External target: vol_mse < 0.0017, surface_mse must NOT regress beyond 0.000296

**Reproduce champion:**
```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --vol-loss-weight 10.0
```

**Context:** nezuko #3177 tests vol-weight=15x/20x WITH asinh+residual-prediction. This PR is the clean control — same weights, no extra features — so we can attribute any improvement to vol-weighting alone vs. the compound effect in nezuko.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT gilbert:

## Bug fix

Found and fixed an `UnboundLocalError` in `train.py:main()` — `primary_metric_key` (a module-level function) was being shadowed by a local string assignment, causing Python to treat it as an unbound local on the earlier function-call line. Fixed by renaming the local variable to `_primary_key` and the function result to `best_val_primary_metric_name`.

## Smoke test (3 epochs, Trial 1 vol-weight=15x)

| Metric | Epoch 1 | Epoch 2 | Epoch 3 |
|---|---|---|---|
| surface_mse | 0.5647 | 0.4254 | 0.3358 |
| vol_mse | 0.5156 | 0.3806 | 0.3202 |

Training is stable, both metrics descending normally. No NaN, no errors.

## Full trials launched

- **Trial 1 (vol-weight=15x):** GPU 0, W&B: gilbert/af-vol-15x-clean
- **Trial 2 (vol-weight=20x):** GPU 1, W&B: gilbert/af-vol-20x-clean

Both running with full 360-min budget, 999 epoch cap. ETA ~6 hours.

**Comment by tcapelle** [comment]
STUDENT gilbert:

## Results

### Trial 1: vol-weight=15x (clean control)

| Metric | Baseline | Trial 1 (val, ep565) | Trial 1 (test, ep565) | Δ vs baseline (val) |
|---|---|---|---|---|
| **surface_mse** | 0.000296 | **0.000352** | 0.001022 | +19% ❌ |
| **vol_mse** | 0.002039 | **0.003402** | 0.004406 | +67% ❌ |
| surface_mse_p | — | 0.001387 | 0.003814 | — |
| vol_mse_Ux | — | 0.006974 | 0.007049 | — |
| vol_mse_Uy | — | 0.004391 | 0.004479 | — |
| vol_mse_nut | — | 0.001630 | 0.005193 | — |
| vol_mse_p | — | 0.000614 | 0.000902 | — |

- **Best val surface_mse:** 0.000352 at epoch 565
- **Vol_mse at best surface epoch:** 0.003402
- **Best val vol_mse observed:** 0.00238 at epoch 763 (not checkpointed — surface_mse was 0.000594 at that epoch)
- **Final epoch:** 764 (360-min timeout)
- **W&B run:** `o048z651`
- **Peak memory:** 43,275 MiB / 97,887 MiB
- **External target vol_mse < 0.0017:** NOT reached (best instantaneous: 0.00238)

**Command:**
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --volume-loss-weight 15.0 \
  --agent gilbert --wandb-name "gilbert/af-vol-15x-clean" \
  --wandb-group af-vol-weight-higher
```

### Trial 2: vol-weight=20x (clean control)

| Metric | Baseline | Trial 2 (val, ep287) | Δ vs baseline |
|---|---|---|---|
| **surface_mse** | 0.000296 | **0.001070** | +261% ❌ |
| **vol_mse** | 0.002039 | **0.006480** | +218% ❌ |

- **Best val surface_mse:** 0.001070 at epoch 287
- **Vol_mse at best epoch:** 0.006480
- **DIVERGED at epoch 314** — catastrophic gradient explosion at cosine lr peak (lr=6e-4). Loss went to NaN, surface_mse flatlined at 0.621, vol_mse at 1.533. EMA model was contaminated (output constant values, component MSEs = 0.0).
- **No test metrics** — process killed after confirming unrecoverable divergence.
- **W&B run:** `pne2413l`

**Command:**
```bash
CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 \
  --volume-loss-weight 20.0 \
  --agent gilbert --wandb-name "gilbert/af-vol-20x-clean" \
  --wandb-group af-vol-weight-higher
```

### What happened

**The hypothesis is rejected.** Neither 15x nor 20x volume loss weighting improves vol_mse below baseline (0.002039), and both significantly regress surface_mse.

**Key findings:**

1. **Surface/volume Pareto tradeoff:** Higher volume weighting pushes vol_mse lower but at the cost of surface_mse. The baseline with 1x weighting sits at a better operating point on this tradeoff curve. At the best checkpoint (ep565), vol_mse is 0.003402 — still 67% above baseline despite 15x volume emphasis.

2. **Vol_mse never reached baseline:** Even the best instantaneous val vol_mse (0.00238 at ep763, late in training) is 17% above baseline. The model would need to continue training well beyond 764 epochs to have any chance of reaching baseline vol_mse, and even then the surface_mse at that point (0.000594) is 2x worse than baseline.

3. **20x is beyond the stability boundary:** Trial 2 diverged catastrophically at epoch 314 when cosine lr hit a peak at 6e-4. The 15x trial showed growing surface spikes at lr peaks (0.00124 at ep570, 0.00138 at ep643) suggesting it's near the stability boundary too.

4. **Vol_mse descent was continuous but slow:** In Trial 1, vol_mse dropped steadily from 0.00487 (ep380) → 0.00301 (ep548) → 0.00238 (ep763). At this rate, reaching baseline vol_mse (0.002039) would require ~1000+ epochs, far beyond practical training budget.

5. **Test metrics show overfitting:** Test surface_mse (0.001022) is 2.9x worse than val surface_mse (0.000352), suggesting the model overfits to validation set, especially for surface pressure.

### Suggested follow-ups

- **Lower volume weights (3x, 5x, 7x):** The Pareto curve suggests moderate weighting might find a sweet spot where vol_mse improves without catastrophic surface regression. The current 1x baseline may be underweighting volume.
- **Cosine annealing with lower lr peak:** Reducing peak lr from 6e-4 to 3e-4 with higher volume weights could avoid the stability boundary that killed Trial 2 and caused growing surface spikes in Trial 1.
- **Multi-objective checkpoint selection:** Select checkpoints based on a combined surface+vol metric rather than surface-only, to capture the best trade-off point.
- **Longer T_max:** Increasing from 50 to 100 would smooth out lr oscillations and reduce the surface spikes at lr peaks that dominate the instability.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3204 because ADVISOR: Closing — both trials worse than baseline. Trial 1 (15x): surface=0.000352 (+19%), vol=0.003402 (+67%). Vol still descending at ep763 but would need ~340 more epochs to reach baseline — impractical. Trial 2 (20x): crashed ep381 from gradient instability at cosine LR peak. Vol-weight=10x is confirmed as the AF sweet spot — 15x and 20x both worsen the surface/volume tradeoff. Interesting observation: best checkpoint is surface-anchored so vol improvements at late epochs (vol=0.002378 at ep763) are hidden behind surface regression. Joint selection metric could be worth exploring separately. Good clean control data, gilbert..
